### ISSUE TITLE
PDO::FETCH_GROUP fetch mode is not working at all

### ISSUE THREAD

**Issue by silwerclaw** [laravel/framework#25176]
- Laravel Version: 5.5.34
- PHP Version: 7.1

### Description:

Pay attention to the `select()` method of `Illuminate\Database\Connection.php`:

**Laravel 5.5**
```
    public function select($query, $bindings = [], $useReadPdo = true)
    {
        return $this->run($query, $bindings, function ($query, $bindings) use ($useReadPdo) {
            if ($this->pretending()) {
                return [];
            }

            // For select statements, we'll simply execute the query and return an array
            // of the database result set. Each element in the array will be a single
            // row from the database table, and will either be an array or objects.
            $statement = $this->prepared($this->getPdoForSelect($useReadPdo)
                              ->prepare($query));

            $this->bindValues($statement, $this->prepareBindings($bindings));

            $statement->execute();

            return $statement->fetchAll();
        });
    }
```

**Laravel 5.2**

```
    public function select($query, $bindings = [], $useReadPdo = true)
    {
        return $this->run($query, $bindings, function ($me, $query, $bindings) use ($useReadPdo) {
            if ($me->pretending()) {
                return [];
            }
            // For select statements, we'll simply execute the query and return an array
            // of the database result set. Each element in the array will be a single
            // row from the database table, and will either be an array or objects.
            $statement = $this->getPdoForSelect($useReadPdo)->prepare($query);
            $statement->execute($me->prepareBindings($bindings));
            $fetchMode = $me->getFetchMode();
            $fetchArgument = $me->getFetchArgument();
            $fetchConstructorArgument = $me->getFetchConstructorArgument();
            if ($fetchMode === PDO::FETCH_CLASS && ! isset($fetchArgument)) {
                $fetchArgument = 'StdClass';
                $fetchConstructorArgument = null;
            }
            return isset($fetchArgument)
                ? $statement->fetchAll($fetchMode, $fetchArgument, $fetchConstructorArgument)
                : $statement->fetchAll($fetchMode);
        });
    }
```

... and this part of code from official 5.4 upgrade, thaty forced us to set fetch mode through some Event thing:

```
Event::listen(StatementPrepared::class, function ($event) {
    $event->statement->setFetchMode(...);
});
```

We can see, that after upgrade `$fetchMode` parameter is now should be set directly to `PDOStatement` object, instead of passing as `fetchAll()` argument. Actually, there is absolutely no difference, if you use primitive fetch modes like `FETCH_ASSOC`, `FETCH_OBJ`. But it does not work at all, when you use fetch modes, targeted for multiple results, like `FETCH_GROUP`.

Unfortunately this is not well documented in php docs but you can check detailed examples of that problem in this stackoverflow thread:
https://stackoverflow.com/questions/20620886/pdostatement-getting-different-results-between-fetchallmode-and-setfetch

**TLDR** of author investigation, crawling native php C code
> The list of constants, supported by setFetchMode (PDOStatement method) does not include the PDO::FETCH_GROUP

It was working for sure before upgrade, but not working with this type of constants after.

### Steps To Reproduce:
```
$fetchMode = PDO::FETCH_GROUP | PDO::FETCH_ASSOC;
\Event::listen(StatementPrepared::class, function ($event) use ($fetchMode) {
        $event->statement->setFetchMode($fetchMode);
});

app('db')->table('fruits')->insert([['name' => 'apple', 'color' => 'green'], ['name' => 'pear', 'color' => 'yellow']]);

$result = app('db')->table('fruits')->get();

dd($result);

```

**Expected result** (results are grouped by the first column value - as it must be using FETCH_GROUP):

```
Illuminate\Support\Collection {
  #items: array:2 [
    "apple" => array:1 [
      0 => array:1 [
        "color" => "green"
      ]
    ]
    "pear" => array:1 [
      0 => array:1 [
        "color" => "yellow"
      ]
    ]
  ]
}
```

**Actual result** (results are not grouped - only FETCH_ASSOC mode applied):

```
Illuminate\Support\Collection {
  #items: array:2 [
    0 => array:2 [
      "name" => "apple"
      "color" => "green"
    ]
    1 => array:2 [
      "name" => "pear"
      "color" => "yellow"
    ]
  ]
}
```

**Comment by staudenmeir** [comment]
Do you have an idea for a fix?

**Comment by silwerclaw** [comment]
I'm currently thinking on that, however with this setFetchMode() callback it's hard to do smth elegant ) Anything that passes `$fetchMode` not into statement but as fetchAll() argument makes the whole idea of the code below useless. Actually almost reverting back to previous version...

```
Event::listen(StatementPrepared::class, function ($event) {
    $event->statement->setFetchMode(...);
});
```

**Comment by mfn** [comment]
My understanding from the whole change back then was to basically discourage alternative fetch modes.

Together with Eloquent apparently they caused all kind of (support) issues.

But on the way the flexibility was sacrificed for regular queries.

**Comment by driesvints** [comment]
I don't think it's meant to work with this mode tbh. If you want this to be supported again it's best to open up an issue in https://github.com/laravel/ideas to get some support first.

**Comment by awesomedeba10** [comment]
Theres another option I find to by pass it
Add env` DB_FETCHMODE=FETCH_ASSOC`
In config/database add for connections.mysql
 `'fetch_mode' => env('DB_FETCHMODE', 'FETCH_ASSOC'),`
In illuminate/datbase/connection.php replace prepared function with
`protected function prepared (PDOStatement $statement){ $config = $this->config; $statement->setFetchMode($config['fetch_mode'] == "FETCH_OBJ" ? 5 : ($config['fetch_mode'] == "FETCH_NUM" ? 3 : 2)); $this->event(new Events\StatementPrepared( $this, $statement )); return $statement; }`

This make default FETCH_ASSOC for your application

Then if you want to change it like before,
add `config(['database.connections.mysql.fetch_mode' => 'FETCH_OBJ']);`
a replacement of `DB::setFetchMode(PDO::FETCH_ASSOC);`
